aten/src/ATen/NestedTensorImpl.cpp

```diff
@@ -219,5 +238,52 @@ const char* NestedTensorImpl::tensorimpl_type_name() const {
   return "NestedTensorImpl";
 }
 
+
+template <typename VariableVersion>
+c10::intrusive_ptr<TensorImpl> NestedTensorImpl::shallow_copy_and_detach_core(
+    VariableVersion&& version_counter,
+    bool allow_tensor_metadata_change) const {
+  if (key_set_.has(DispatchKey::Python) &&
+      !c10::impl::tls_is_dispatch_key_excluded(DispatchKey::Python)) {
+    auto r = pyobj_interpreter_.load(std::memory_order_acquire)->detach(this);
+    if (r) {
+      r->set_version_counter(std::forward<VariableVersion>(version_counter));
+      r->set_allow_tensor_metadata_change(allow_tensor_metadata_change);
+      return r;
+    }
+    // otherwise just copy the TensorImpl and not the PyObject.  Since
+    // the interpreter is dead no one can call us out on it
+  }
+  auto impl = c10::make_intrusive<NestedTensorImpl>(
+      buffer_size_,
+      storage_,
+      key_set_,
+      data_type_,
+      nested_size_tensor_,
+      nested_stride_tensor_,
+      offsets_);
+
+      copy_tensor_metadata(
+          /*src_impl=*/this,
+          /*dest_impl=*/impl.get(),
+          /*version_counter=*/std::forward<VariableVersion>(version_counter),
+          /*allow_tensor_metadata_change=*/allow_tensor_metadata_change);
```

Comment:
You need to manually copy the NestedTensorImpl specific fields here right?

Reply:
I created that new constructor which is used on line 257. Makes sure to copy over all nested specific meta data when creating the new tensor impl.  
`copy_tensor_metadata` will then copy over the version version_counter. 
And call:
https://github.com/pytorch/pytorch/blob/cd5efc6f082c81fd40712127638931b9e2e5ee69/c10/core/TensorImpl.cpp#L623

https://github.com/pytorch/pytorch/blob/cd5efc6f082c81fd40712127638931b9e2e5ee69/c10/core/TensorImpl.cpp#L596

There is some repeated work going on here. Since the above constructor populates all the nested specific fields. I think have this constructor is good since it is the most specific.